Support older versions of Consul in sync catalog

lkysow opened the pull request:
If namespaces are not enabled, use a Consul catalog API that is
available in older Consul versions instead of the API that's only
available in Consul 1.7+.

Design decisions:
* I've only implemented the adapter for the `NodeServiceList` endpoint rather than for the whole consul client API. This means `Syncer` needs both `ConsulClient` and `ConsulNodeServiceListClient` but it reduces the amount of code required vs needing to wrap the whole API of `ConsulClient`.
* Previously (before namespaces) the code used the `/catalog/services` endpoint to return just service names and then it filtered on those with the tag `k8s`. It never checked if those services were registered to the `k8s-sync` node. Rather than going back to that API, I've switched it to the `/catalog/node/:node_id` endpoint. I made this change because it matches what the namespaces code is doing by hitting the `/catalog/node-services/:node_id` endpoint. This makes the adapter easier to understand since it's adapting "list services for a node" instead of "list services for a node if namespaces, else list all services regardless of node". It is also backwards compatible because the only node the sync process ever registered services on is `k8s-sync`.
* Rather than having `ConsulClient` be a field in the adapting client, I'm making it part of the function signature. I made this decision because I liked that the Syncer was using the same underlying Consul client for all its calls. If it redefined its own `ConsulClient` field (it doesn't right now), then the clients would be out of sync if the adapting client had its own.
* I've changed the implementation of the namespaces-aware client to add a filter on tags. This filtering was originally being done in code but it seems better to use the proper API capabilities for this.

Manual Testing:
1. Consul OSS 1.6.3 - Use code before change, create a service manually with tag "k8s", see that it **does not** get deleted. This is because the catalog list call is failing and so we never populate the list of services to delete.
1. Consul OSS 1.6.3 - Use code after change, create a service manually with tag "k8s" on node k8s-sync, see that is deleted (proving that the bug is fixed)
1. Consul OSS 1.7.0 - Run with namespaces disabled, manually create a service with tag k8s on node k8s-sync, see that it is deleted.
1. Consul Enterprise 1.7.0 - Run with namespaces enabled syncing to one namespace. Change config to mirroring. See that services in old namespace are deleted.

lkysow commented on .circleci/config.yml:
```diff
@@ -6,7 +6,7 @@ executors:
     environment:
       - TEST_RESULTS: /tmp/test-results # path to where test results are saved
-      - CONSUL_VERSION: 1.7.0 # Consul's OSS version to use in tests
+      - CONSUL_VERSION: 1.6.3 # Consul's OSS version to use in tests
```
todo: we probably want to switch this to 1.7.0 but I have this here now to prove that the tests pass with an older consul version.

adilyse commented on catalog/to-consul/consul_node_services_client_ent_test.go:
```diff
@@ -0,0 +1,343 @@
+			},
+			Exp: nil,
+		},
+		"service on k8s node but without tag": {
```
It would be good to have a test like this, only with a tag or two that don't match "k8s". That will make sure we don't accidentally delete services from a different sync process that's using a different tag.

adilyse commented on catalog/to-consul/consul_node_services_client_ent_test.go:
```diff
@@ -0,0 +1,343 @@
+		"service on k8s node but without tag": {
+			ConsulServices: []api.CatalogRegistration{
+				{
+					Node:    "k8s-sync",
```
These tests could reference the `ConsulSyncNodeName` so that if it gets updated, these also get updated.

adilyse commented on catalog/to-consul/consul_node_services_client_ent_test.go:
```diff
@@ -0,0 +1,343 @@
+			require.NoError(err)
+			for _, registration := range c.ConsulServices {
+				if registration.Service.Namespace != "" && registration.Service.Namespace != "default" {
+					_, _, err = consulClient.Namespaces().Create(&api.Namespace{
```
For tests with multiple services in a non-default namespace, does this create that namespace multiple times?

lkysow replied:
It makes the API call multiple times but no error is returned if the ns already exists

adilyse commented on catalog/to-consul/consul_node_services_client_test.go:
```diff
@@ -0,0 +1,168 @@
+			},
+			Exp: nil,
+		},
+		"service on k8s node but without tag": {
```
Similar to the ent tests, it would be good to duplicate this test but put non-`k8s` tags on it.

lkysow commented on catalog/to-consul/consul_node_services_client.go:
```diff
@@ -0,0 +1,113 @@
+		return nil, meta, nil
+	}
+	var svcs []ConsulService
+	// seenServices is used to ensure the svcs list is unique.
```
Yes it returns items by their service ids, not the service names. So if there's multiple instances of a service the service name will be duplicated.

lkysow commented on catalog/to-consul/consul_node_services_client.go:
```diff
@@ -0,0 +1,113 @@
+					svcs = append(svcs, ConsulService{
+						// If namespaces are not enabled we use empty
+						// string.
+						Namespace: "",
```
Yes but I thought it best to be explicit here.

lkysow commented on catalog/to-consul/consul_node_services_client.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// ConsulNodeServicesClient is used to query for node services.
+type ConsulNodeServicesClient interface {
```
I'm not sure about removing Consul here. The `catalog` package contains both consul and kubernetes code and it's conceivable that there could be a NodeServicesClient for Kubernetes. I'd like to leave it as is unless you feel strongly about this change.

adilyse replied:
Maybe the middle ground here is to keep the `Consul` in the interface name but remove it from the implementations. Then it's `PreNamespacesNodeServicesClient` that implements the `ConsulNodeServicesClient` interface. That should still give a clear indication that it's specific to Consul.

lkysow commented on catalog/to-consul/consul_node_services_client.go:
```diff
@@ -0,0 +1,113 @@
+
+// ConsulPreOnePointSevenNodeServicesClient implements ConsulNodeServicesClient
+// for Consul < 1.7 which does not support namespaces.
+type ConsulPreOnePointSevenNodeServicesClient struct{}
```
I've changed `PreOnePointSeven` to `PreNamespaces`. I think it's important to keep the `Consul` prefix for the reasons stated above.

I'm of two minds on keeping `NodeServicesClient` suffix: I like that it matches the interfaces it's implementing and I think it makes it clear what it is. If we drop the `Client` then it sounds like the object represents node services themselves. The downside is that it's more verbose.

Another approach here would be to put this code in a new package, maybe `nodeservices`. Then we could use: `Client`, `PreNamespacesClient` and `NamespacesClient`.

lkysow commented on catalog/to-consul/consul_node_services_client.go:
```diff
@@ -0,0 +1,113 @@
+	// registered to the Consul node with nodeName. opts is used as the
+	// query options in the API call to consul. It returns the list of services
+	// (not service instances) and the query meta from the API call.
+	NodeServicesWithTag(client *api.Client, tag string, nodeName string, opts *api.QueryOptions) ([]ConsulService, *api.QueryMeta, error)
```
Good point about the pointer meaning there's no change. Done.

lkysow commented:
* updated to make the consul client a field in each client
* updated names but also discussed moving this to its own package which would make naming easier, what do you think?
* The difference is that previously we would delete any service tagged k8s that we didn't create. Now we'll only delete that service if it existed on the k8s-sync node. I think we should definitely note this in the changelog but it seems like the previous behaviour was buggy. It makes sense that we clean up services *we've* registered when our config changes. We know that all the services we register are on the k8s-sync node. So previously I think we were overstepping our bounds.

lkysow commented:
Okay, feel free to rebase that last "code review updates" commit.
